### DISCUSSION TITLE
tutorials/shaders/screen-reading_shaders

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#151] [User-contributed notes]
# tutorials/shaders/screen-reading_shaders

Introduction: It is often desired to make a shader that reads from the same screen to which it's writing. 3D APIs, such as OpenGL or DirectX, make this very difficult because of internal hardware l...

https://docs.godotengine.org/en/stable/tutorials/shaders/screen-reading_shaders.html

**Comment by DeveCout** [comment]
I can't find the test project to test the lens effect. A link is missing or somethings ?

**Reply by Calinou** [reply]
I don't think this test project was released anywhere, unfortunately. It would have to be recreated from scratch at this point.

**Comment by ramblingstranger** [comment]
A significant limitation with 3D transparent objects is mentioned in this doc:

> In 3D, the screen is copied after the opaque geometry pass, but before the transparent geometry pass, so transparent objects will not be captured in the screen texture.

This affects not just the screen texture but also the depth and normal-roughness textures.

There are a couple ways to get around this so that you can do post-processing effects with transparent objects:

1. If you are only using the screen texture (not depth or normal-roughness), use a 2D shader. When you access the screen texture in a 2D shader, transparent objects will show up. (I believe this is because the 3D rendering is all already finished.) This doesn't work for the depth or normal-roughness textures, because those are not available in 2D shaders.

2. Use a [compositor effect](https://docs.godotengine.org/en/stable/tutorials/rendering/compositor.html). This is a lower-level solution which requires writing GLSL and passing data like in a compute shader. Note that you will need to set transparent objects to "depth draw always" in order for them to show up in the depth buffer. And currently, there is no way to get transparent objects to show up in the normal-roughness texture.

3. You can use a setup with multiple cameras and viewports to capture and pass textures manually. This is kind of complicated to set up, but it ends up giving you more control than a compositor effect. You can access color, depth, normal, or roughness data, or you can pass your own custom data. I have an example project here: https://github.com/ramblingstranger/custom-screen-buffers

**Comment by SephReed** [comment]
Can anyone give a minimum viable example of a 3d scene with screen space shader?  It could simply be a greyscale shader, but it would help tremendously.

**Reply by nc5432** [reply]
A bit late, but I just set up an example project here: https://github.com/nc5432/Godot-Screen-Space-Shader-Example

**Comment by mucketban** [comment]
hint_depth_texture isn't supported on canvas_item shaders, what's the best way to access the depth texture for a screen-space shader?

**Reply by Calinou** [reply]
2D rendering in Godot does not have a depth buffer, so there is no `depth_texture` to access. See https://github.com/godotengine/godot-proposals/issues/783.

In 3D, use a screen-aligned quad with a `spatial` shader as described in [Advanced post-processing](https://docs.godotengine.org/en/stable/tutorials/shaders/advanced_postprocessing.html).

**Comment by snail-with-tea** [comment]
Can hint_normal_roughness be enabled optionally? since both webgl2 and most of mobile platforms support multiple render targets

**Reply by snail-with-tea** [reply]
At least this should be faster then reconstructing normals from depth, right?
Really want to recreate t3ssel8r style on web, and i know it is possible since [this example](https://threejs.org/examples/#webgl_postprocessing_pixel) exists

**Reply by Calinou** [reply]
`hint_normal_roughness` isn't implemented in the Compatibility renderer in the first place (not even on desktop platforms), so this isn't feasible.

**Comment by m-radzikowski** [comment]
While I can't find it in documentation anywhere, it seems that shaders on separate CanvasLayers receive separate full-screen copies. So the shader on a higher layer will use the results of screen-reading shaders on lower layers, similarly to using BackBufferCopy or Viewports.
